plugins/flexibleengine/src/main/java/org/eclipse/xpanse/plugins/flexibleengine/monitor/MetricsService.java

```diff
@@ -204,33 +190,46 @@ public List<Metric> getMetricsForResource(ResourceMetricRequest resourceMetricRe
-                                        resourceMetricRequest.isOnlyLastKnownMetric());
-
-                        if (!CollectionUtils.isEmpty(metric.getMetrics())) {
-                            metrics.add(metric);
-                            flexibleEngineMonitorCache.set(deployResource.getResourceId(), metric);
-                        } else {
-                            List<Metric> metricCache =
-                                    flexibleEngineMonitorCache.get(deployResource.getResourceId(),
-                                            metricInfoList.getMetricName());
-                            metrics.addAll(metricCache);
-                        }
+            try {
+                Project project =
+                        queryProjectInfo(credential, projectQueryUrl);
+                if (Objects.nonNull(project) && StringUtils.isNotBlank(project.getId())) {
+                    String listMetricsUrl =
+                            flexibleEngineMonitorConverter.buildListMetricsUrl(deployResource,
+                                    project.getId());
+                    ListMetricsResponse listMetricsResponse =
+                            queryMetricItemList(credential, listMetricsUrl);
+                    List<MetricInfoList> targetMetrics =
+                            getTargetMetricInfoList(deployResource, listMetricsResponse,
+                                    resourceType);
+                    for (MetricInfoList metricInfoList : targetMetrics) {
+                        String url = flexibleEngineMonitorConverter.buildMonitorMetricUrl(
+                                resourceMetricRequest, project.getId(), metricInfoList);
+                        if (StringUtils.isNotBlank(url)) {
+                            ShowMetricDataResponse response = queryMetricData(credential, url);
+                            Metric metric =
+                                    flexibleEngineMonitorConverter.convertResponseToMetric(
+                                            deployResource, response, metricInfoList,
+                                            resourceMetricRequest.isOnlyLastKnownMetric());
+
+                            if (!CollectionUtils.isEmpty(metric.getMetrics())) {
+                                metrics.add(metric);
+                                flexibleEngineMonitorCache.set(deployResource.getResourceId(),
+                                        metric);
+                            } else {
+                                List<Metric> metricCache =
+                                        flexibleEngineMonitorCache.get(
+                                                deployResource.getResourceId(),
+                                                metricInfoList.getMetricName());
+                                metrics.addAll(metricCache);
+                            }
 
+                        }
                     }
                 }
+            } catch (Exception e) {
```

Comment:
@baixinsui If there is an exception from backend, its better we simply return it to client.  Because the client must differentiate between "no metrics found" and "error while getting metrics".
1. Try-catch only around the backend API calls. 
2. Define a new exception called - For example - BackendApiCallFailed
3. Map this to HTTP 502 (which is bad gateway). 

Ideally, if we know what exception can be expected here, then we must avoid catching "Exception" here and instead specific exceptions must be catched.

Reply:
Ok.